Turbulence.ModelGenerator/Preprocessing.cs

```diff
@@ -0,0 +1,136 @@
+using System.Globalization;
+using System.Text;
+using System.Text.RegularExpressions;
+
+namespace Turbulence.ModelGenerator;
+
+public static class Preprocessing
+{
+    /// <summary>
+    /// Extract relevant tables from .md files.
+    /// </summary>
+    public static async Task ExtractTables(Uri inPath, Uri outPath)
+    {
+        // List to keep track of written files, so we can make sure there are no duplicates
+        List<string> writtenFiles = new();
+
+        foreach (string file in Directory.EnumerateFiles(inPath.AbsolutePath, "*.*", SearchOption.AllDirectories))
+        {
+            using StreamReader reader = new(file);
+
+            // Keep track of the last ## header found, because it will be used to name JSON Params, JSON/Form and Query String Params tables
+            string? lastH2 = null;
+
+            string? line;
+            while ((line = await reader.ReadLineAsync()) != null)
+            {
+                if (line.StartsWith("## "))
+                {
+                    lastH2 = line;
+                }
+                
+                // Based on the assumption that every table has H6 above it, which it should according to the README
+                if (!line.StartsWith("###### ")) continue;
+
+                string? name;
+                
+                /* language=RegEx */
+                const string paramsPattern = @"###### (.*?)(Query String Params|JSON Params|JSON/Form Params|Response Body)$";
+                var match = Regex.Match(line, paramsPattern);
+                if (match.Success && lastH2 != null)
+                {
+                    // Use the last header title instead of [...] Params
+                    name = Regex.Match(lastH2, @"## (.*) %.*").Groups[1].Value;
+
+                    if (line.Contains("Query String Params"))
+                    {
+                        name += "QueryParams";
+                    }
+                    else if (line.Contains("Response Body"))
+                    {
+                        name += "Response";
+                    }
+                    else
+                    {
+                        name += "Params";
+                    }
+                }
+                else
+                {
+                    // Markdown heading to pretty class name
+                    /* language=RegEx */
+                    const string pattern = @"###### (.*?)( Structure| Fields)?$";
```

Comment:
Wouldn't it maybe be better to compile the regex beforehand? Not sure if it caches them